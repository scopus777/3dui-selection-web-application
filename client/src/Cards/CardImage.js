import React, {Component} from "react";

/*
* This Component is used to display an image in a card environment.
* Props:
*   - image : The url where the image is located
*   - technique : The technique this image stems from
* */
class CardImage extends Component {
    render() {
        return (
            <img className="card-img-top card-image-cover"
                 src={this.props.image.image}
                 alt={this.props.technique.name + " example image."}/>
        )
    }
}

export default CardImage;
