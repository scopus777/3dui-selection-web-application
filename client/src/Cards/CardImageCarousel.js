import {Carousel} from "react-bootstrap";
import React, {Component} from "react";
import CardImage from "./CardImage";

/* Required to load in a FontAwesome Icon in React*/
const FontAwesome = require('react-fontawesome');

/* This Component creates a Carousel component with all the images. To do this it needs to know which images belong to the passed over technique. Therefore it connects to the server first
* Props:
*   - technique: The technique which images should be displayed
* */
class CardImageCarousel extends Component {

    /*Initialize that we will be using a list for the images in the State object*/
    constructor(props) {
        super(props);
        this.state = {
            images: []
        }
    }

    /*After the component mounts we fetch the data*/
    componentDidMount() {
        this.fetchData();
    }

    /*After a prop has changed this function is called*/
    componentDidUpdate(prevProps) {
        //Check to see if it was the technique id prop that was changed
        //This has the reason that otherwise the image data would not be fetched if the technique is set later
        if (this.props.technique.id !== prevProps.technique.id) {
            this.fetchData();
        }
    }

    /*This function fetches the urls from the server for the images*/
    fetchData(){
        fetch('/techniques/' + this.props.technique.id + '/images')
            .then(res => res.json())
            .then(data => this.setState({images: data}));
    }

    render() {
        let imageCarousel = "";
        //If we have more then 1 image then we have to display a Carousel to go through the different images.
        if (this.state.images.length > 1) {
            //We have to change the icons for previous and next, so that they work on dark and bright backgrounds
            imageCarousel =
                <Carousel prevIcon={<FontAwesome name='angle-left' size='2x'/>} nextIcon={<FontAwesome name='angle-right' size='2x'/>} interval={10000}>
                    {
                        this.state.images.map(
                            image =>
                                <Carousel.Item key={image.id}>
                                    <CardImage key={image.id} image={image} technique={this.props.technique}/>
                                </Carousel.Item>
                        )
                    }
                </Carousel>
        } else if (this.state.images.length === 1) { //Otherwise we can just display the image
            imageCarousel = <CardImage image={this.state.images[0]} technique={this.props.technique}/>
        }
        else {
            let image = {image: "/images/no_image.png"};
            imageCarousel = <CardImage image={image} technique={this.props.technique}/>
        }

        return (imageCarousel)
    };
}


export default CardImageCarousel;
