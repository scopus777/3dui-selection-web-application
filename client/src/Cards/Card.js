import React, {Component} from 'react';
import {Link} from 'react-router-dom';
import {Card as CardBootstrap} from 'react-bootstrap';
import CardImageCarousel from "./CardImageCarousel";

/*
* This Component displays a bootstrap Card Component with all the values of a specific technique
* These values are passed over via props
* */
class Card extends Component {

    render() {
        let view = [];
        let addInfo = [];
        // show metrics if the card is used in the sort view
        if (this.props.showMeasurement) {
            if (this.props.technique.addInfo !== undefined)
                addInfo = this.props.technique.addInfo.split(";");
            for (let i = 0; i < addInfo.length; i++) {
                if (addInfo[i].startsWith("<b>")) {
                    view.push(
                        <p key={"card_" + i} className="card-text mb-1">
                            <small className="text-muted">
                                <b>{addInfo[i].replace("<b>", "")} </b>
                            </small>
                        </p>
                    )
                } else {
                    view.push(
                        <p key={"card_" + i} className="card-text mb-1">
                            <small className="text-muted">
                                {addInfo[i]}
                            </small>
                        </p>
                    )

                }
            }
        }

        return (
            //These Bootstrap classes define 4 columns on large screens, 3 columns on medium screens, 2 columns on small screens, 1 column on very small screens
            <div className="col-xs-12 col-sm-4 col-md-3 col-lg-3 col-xl-2 p-2">
                {/*Generate a Bootstrap Card with a shadow, the text centered and with the same height as the other cards*/}
                <CardBootstrap className="shadow-sm bg-light text-center  h-100">
                    {/*The card has an image on top, which is handled by the CardImageCarousel Component*/}
                    <CardImageCarousel technique={this.props.technique}/>
                    {/*The card's body has a title and a short description*/}
                    <CardBootstrap.Body className="d-flex flex-column">
                        <CardBootstrap.Title><b>{this.props.technique.name}</b></CardBootstrap.Title>
                        <CardBootstrap.Text>{this.props.technique.description_short}</CardBootstrap.Text>
                        <div className="mt-auto">
                            {/*Because we are using react-router we have to use the Link component instead of an <a>*/}
                            <Link to={"/detail/" + this.props.technique.id} className="btn btn-primary">
                                Details
                            </Link>
                        </div>
                    </CardBootstrap.Body>
                    {/*In the lower portion the metrics of the technique is displayed*/}
                    {
                        view
                    }
                </CardBootstrap>
            </div>
        );
    }
}

export default Card;
