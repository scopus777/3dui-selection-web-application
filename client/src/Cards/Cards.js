import React, {Component} from 'react';
import Card from "./Card";
import CardGroup from "react-bootstrap/CardGroup";

/*
* This Component creates a list of cards.
* Props:
*   - techniques : All the techniques that should be displayed
* */
class Cards extends Component {

    render() {
        return (
            <CardGroup>
                {
                    //Go over every technique and create a Card component for that technique
                    this.props.techniques.map(technique =>
                        <Card key={technique.id} technique={technique}
                                       confidence={technique.confidence} confidence_metrics={technique.confidence_metrics} showMeasurement={this.props.showMeasurement}/>
                    )
                }
            </CardGroup>
        );
    }
}

export default Cards;

