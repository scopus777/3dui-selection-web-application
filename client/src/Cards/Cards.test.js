import React from 'react';
import Cards from './Cards';
import {create} from 'react-test-renderer'
import Card from "./Card";
import CardImage from "./CardImage";
import CardImageCarousel from "./CardImageCarousel";
import {MemoryRouter} from "react-router-dom";

const testTechnique = {
    id: 1,
    name: "Simple Virtual Hand",
    description: "Creates a virtual representation of the real hand to allow a direct and natural interaction",
};

describe('Test Card Component Snapshots', () => {
    test('testing CardImage snapshot', () => {
        let tree = create(<CardImage image="#" technique={testTechnique}/>);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('testing CardImageCarousel snapshot', () => {
        let tree = create(<CardImageCarousel technique={testTechnique}/>);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('testing Card snapshot', () => {
        let tree = create(
            <MemoryRouter>
                <Card technique={testTechnique}/>
            </MemoryRouter>);
        expect(tree.toJSON()).toMatchSnapshot();
    });

    test('testing Cards snapshot', () => {
        let tree = create(
            <MemoryRouter>
                <Cards techniques={[
                    testTechnique,
                    testTechnique,
                    testTechnique
                ]}/>
            </MemoryRouter>);
        expect(tree.toJSON()).toMatchSnapshot();
    });
});

