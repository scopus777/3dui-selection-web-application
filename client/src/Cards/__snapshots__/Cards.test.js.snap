// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Card Component Snapshots testing Card snapshot 1`] = `
<div
  className="col-xs-12 col-sm-6 col-md-6 col-lg-4 col-xl-3 p-2"
>
  <div
    className="shadow-sm bg-light text-center h-100 card"
  >
    
    <div
      className="d-flex flex-column card-body"
    >
      <div
        className="card-title h5"
      >
        <b>
          Simple Virtual Hand
        </b>
      </div>
      <p
        className="card-text"
      >
        Creates a virtual representation of the real hand to allow a direct and natural interaction
      </p>
      <div
        className="mt-auto"
      >
        <a
          className="btn btn-primary"
          href="/detail/1"
          onClick={[Function]}
        >
          Details
        </a>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test Card Component Snapshots testing CardImage snapshot 1`] = `
<img
  alt="Simple Virtual Hand example image."
  className="card-img-top card-image-cover"
/>
`;

exports[`Test Card Component Snapshots testing CardImageCarousel snapshot 1`] = `""`;

exports[`Test Card Component Snapshots testing Cards snapshot 1`] = `
<div
  className="card-group"
>
  <div
    className="col-xs-12 col-sm-6 col-md-6 col-lg-4 col-xl-3 p-2"
  >
    <div
      className="shadow-sm bg-light text-center h-100 card"
    >
      
      <div
        className="d-flex flex-column card-body"
      >
        <div
          className="card-title h5"
        >
          <b>
            Simple Virtual Hand
          </b>
        </div>
        <p
          className="card-text"
        >
          Creates a virtual representation of the real hand to allow a direct and natural interaction
        </p>
        <div
          className="mt-auto"
        >
          <a
            className="btn btn-primary"
            href="/detail/1"
            onClick={[Function]}
          >
            Details
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="col-xs-12 col-sm-6 col-md-6 col-lg-4 col-xl-3 p-2"
  >
    <div
      className="shadow-sm bg-light text-center h-100 card"
    >
      
      <div
        className="d-flex flex-column card-body"
      >
        <div
          className="card-title h5"
        >
          <b>
            Simple Virtual Hand
          </b>
        </div>
        <p
          className="card-text"
        >
          Creates a virtual representation of the real hand to allow a direct and natural interaction
        </p>
        <div
          className="mt-auto"
        >
          <a
            className="btn btn-primary"
            href="/detail/1"
            onClick={[Function]}
          >
            Details
          </a>
        </div>
      </div>
    </div>
  </div>
  <div
    className="col-xs-12 col-sm-6 col-md-6 col-lg-4 col-xl-3 p-2"
  >
    <div
      className="shadow-sm bg-light text-center h-100 card"
    >
      
      <div
        className="d-flex flex-column card-body"
      >
        <div
          className="card-title h5"
        >
          <b>
            Simple Virtual Hand
          </b>
        </div>
        <p
          className="card-text"
        >
          Creates a virtual representation of the real hand to allow a direct and natural interaction
        </p>
        <div
          className="mt-auto"
        >
          <a
            className="btn btn-primary"
            href="/detail/1"
            onClick={[Function]}
          >
            Details
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
